tool/tool.c

```diff
@@ -0,0 +1,266 @@
+  printf("       memory inefficient boot sector patch if it's loaded.\n");
+  printf("\n");
+
+  return 1;
+}
+
+int parse_arguments(int argc, char **argv, config_t *config) {
+  int parsed = 0;
+  int c;
+
+  // See: https://open-watcom.github.io/open-watcom-v2-wikidocs/clib.html#getopt
+  while((c = getopt(argc, argv, "fd:rpwnt")) != -1) {
+    parsed++;
+    switch(c) {
+      case 'f':
+        printf("Forcing the issue.\n");
+        break;
+
+      case 'd':
+        // Drive can only be a byte value.
+        config->drive = atoi(optarg) & 0xFF;
+        break;
+
+      case 'r':
+        // Read the boot sector and dump it as raw hex
+        config->dump = true;
+        break;
+
+      case 'n':
+        config->ignore_partition_table = true;
+        break;
+      
+      case 'p':
+        config->output_partition_table = true;
+        break;
+
+      case 'w':
+        config->write = true;
+        break;
+
+      case 't':
+        config->load_tsr = true;
+        break;
+
+      case ':':
+        printf("%c requires and argument.\n", optopt);
+        return 1;
+        break;
+
+      case '?':
+      default:
+        printf("Unknown argument '%c'\n", optopt);
+        return 1; 
+    }
+  }
+
+  return 0;
+}
+
+/**
+ * Check to see if the boot in the buffer has our signature.
+ */
+bool is_drive_patched(mbr_t *sector_buf) {
+  return sector_buf->mbr.signature == TSR_SIGNATURE;
+}
+
+/**
+ * Updates the on disk MBR with our active one preserving the
+ * existing partition table.
+ */
+uint8_t write_mbr(config_t *config, mbr_t *original) {
+  uint8_t status;
+  mbr_t write_buf;
+
+  if (sizeof(mbr_t) != boot_boot_bin_len) {
+    printf("FATAL ERROR: MBR structure and Boot Sector are not the same size!!!\n");
+    printf("MBR Struct %u Boot Sector %u\n", sizeof(mbr_t), boot_boot_bin_len);
+  }
+  
+  memcpy(write_buf.buffer, boot_boot_bin, sizeof(mbr_t));
+ 
+  // Copy the partition table over from the original boot sector.
+  if (!config->ignore_partition_table) {
+    printf("Found existing partition table, copying.\n");
+    memcpy(
+        write_buf.mbr.partition_table,
+        (original->mbr.partition_table),
+        sizeof(partition_entry_t) * 4);
+
+  }
+
+  printf("Using partition table:\n");
+  print_partition_table(write_buf.mbr.partition_table);
+
+  // Give users a chance to back out before nuking their drive. 
+  nasty_warning();
+
+  printf("Writing active boot sector...\n");
+
+  // Actually write the boot sector. 
+  status = write_boot_sector(config->drive, &write_buf);
+
+  if (status) {
+    printf("Error writing boot sector %02X -> %s\n", status, translate_error(status));
+  }
+
+  printf("Done!\n");
+
+  printf("A reboot is necessary for these changes to take effect.\n");
+
+  return status;
+}
+ 
+int main(int argc, char **argv) {
+  far_ptr_t mbr_expected;
+  config_t config;
+  mbr_t sector_buf;
+  uint8_t status;
+
+  //tsr_data_paragraphs = (0x100 + 15 + (uint16_t)&tsr_data_paragraphs) >> 4;
```

Comment:
I've tried both of these and neither seems to work quite right. .. Either I'm not keeping the data in memory or the interrupt vector is pointing at the wrong spot.